In [224]:
import numpy as np
import pandas as pd
from itertools import combinations
from math import pi, e

You and your friend are standing in front of a magician and he has a basket with 6 white,
4 red and 5 blue balls, each time the magician draws 3 balls and asks you to predict
what would be the color of the balls. If you predict correctly the magician loses the
game and you win. Since you have become a statistician after doing the ML assignment,
what should be the optimal color scheme of the balls that you would propose in order
to win the game?

In [225]:
white = ['W' for x in range(6)]
red = ['R' for x in range(4)]
blue = ['B' for x in range(5)]

balls = white
balls.extend(red)
balls.extend(blue)

balls

['W', 'W', 'W', 'W', 'W', 'W', 'R', 'R', 'R', 'R', 'B', 'B', 'B', 'B', 'B']

In [226]:
a = list(combinations(balls, 3))
uniqueComb = list(set(a))
l = [a.count(j) for i, j in enumerate(uniqueComb)]
        
print('Combination:', uniqueComb[l.index(max(l))], '\n\nProbability:', '%.2f' % (max(l)/len(a)))

Combination: ('W', 'R', 'B') 

Probability: 0.26


Write a function that takes input from user and determines the input parameters belong
which class (using Naïve assumptions) (use the provided dataset) . Compare your results
with GaussianNB() built in function

In [227]:
df = pd.read_csv('iris.csv')
df

sepal.length  sepal.width  petal.length  petal.width    variety
0             5.1          3.5           1.4          0.2     Setosa
1             4.9          3.0           1.4          0.2     Setosa
2             4.7          3.2           1.3          0.2     Setosa
3             4.6          3.1           1.5          0.2     Setosa
4             5.0          3.6           1.4          0.2     Setosa
..            ...          ...           ...          ...        ...
145           6.7          3.0           5.2          2.3  Virginica
146           6.3          2.5           5.0          1.9  Virginica
147           6.5          3.0           5.2          2.0  Virginica
148           6.2          3.4           5.4          2.3  Virginica
149           5.9          3.0           5.1          1.8  Virginica

[150 rows x 5 columns]

In [228]:
%pylab inline 

%pylab is deprecated, use %matplotlib inline and import the required libraries.
Populating the interactive namespace from numpy and matplotlib


/opt/homebrew/Cellar/jupyterlab/3.4.5/libexec/lib/python3.10/site-packages/IPython/core/magics/pylab.py:162: UserWarning: pylab import has clobbered these variables: ['pi', 'e']
`%matplotlib` prevents importing * from pylab and numpy
  warn("pylab import has clobbered these variables: %s"  % clobbered +



$$f_{\mathbf x}(x_1,\ldots,x_k) = \frac{1}{\sqrt{(2\pi)^k|\boldsymbol\Sigma|}} \exp\left(-\frac{1}{2}({\mathbf x}-{\boldsymbol\mu})^\mathrm{T}{\boldsymbol\Sigma}^{-1}({\mathbf x}-{\boldsymbol\mu}) \right)$$

In [231]:
def GuassianNB(X, df, d, meu):
    covMatrix = np.zeros((d,d))
    for i in range(d):
        for j in range(d):
            if i == j:
                covMatrix[i][j] = np.array(df.cov()).diagonal()[i]
    determinant = np.linalg.det(covMatrix)
    a = ((X - meu).transpose())
    b = np.linalg.inv(covMatrix)
    c = X - meu
    
    first = (1 / ((2*(pi**(d))*determinant)**(1/2)))
    second = e**((-0.5)*((a.dot(b)).dot(c)))
    
    return first*second

GuassianNB(np.array([5.8, 4.0, 1.2, 0.2], dtype='float32'), df, 4, mean)*probSetosa

TypeError: unsupported operand type(s) for -: 'float' and 'function'

In [ ]:
setosa = np.array(df.drop(df.columns[-1], axis = 1))
versicolor = setosa[50:]
setosa = setosa[:50]
virginica = versicolor[:50]

# mean = versicolor.mean(axis=0)
mean = setosa.mean(axis=0)
# mean = virginica.mean(axis=0)
mean

In [ ]:
probSetosa = len(setosa)/len(np.array(df))
probVersicolor = len(versicolor)/len(np.array(df))
probVirginia = len(virginica)/len(np.array(df))

probSetosa

In [ ]:
np.array(df)

### Bonus

A friend of yours has just opened a pizza shop and he doesn’t want to you come there
fearing you would eat for free. Somehow you sneak in the shop and you order an 8 inch
pizza , your friend tells you that he doesn’t have an 8 inch pizza but he can give you two
5 inch ones at the same price with the exact same things. As a statistician, should you
be taking that deal? Is that beneficial for you or not? Give statistical reasoning.

In [1]:
from math import pi

a8 = pi*(8/2)**2
a5 = pi*(5/2)**2

a8, a5*2

(50.26548245743669, 39.269908169872416)

In [8]:
print(int(a5*2/a8*100), '\b%')

78 %


Bad deal as two 5 inch pizzas is 78% more of the 8 inch one